File: 2019/Thang03_2019/16_03_2019/dance.pas
const maxn = 201; maxp = 200*200+1; fi = 'dance.inp'; fo = 'dance.out';
  INF = 100;
var a : array[0..1, 0..maxp] of int64;
  l : array[0..maxn] of longint;
  p : array[0..maxn] of int64;
  n, tests, t, i, j, id, pos, min_pos, max_pos : longint;
  ret : int64;

function max(a, b : int64) : int64;
begin
  if a > b then exit(a) else exit(b);
end;

begin
  assign(input, fi); reset(input);
  assign(output, fo); rewrite(output);
  readln(tests);

  for t := 1 to tests do begin
    readln(n);
    for i := 0 to n - 1 do read(l[i]);
    for i := 0 to n - 1 do read(p[i]);

    a[0][0] := 0; id := 0;
    pos := 0; max_pos := 0;
    for i := 0 to n - 1 do begin
      id := 1 - id;
      if l[i] > 0 then max_pos := pos + l[i];
      for j := 0 to max_pos do a[id][j] := -INF;
      for j := 0 to pos do a[id][j] := a[1 - id][j];
      min_pos := pos + l[i];
      if min_pos > max_pos then min_pos := max_pos;
      for j := max(0, l[i]) to min_pos do
        a[id][j] := max(a[id][j], p[i] + a[1 - id][j - l[i]]);

      pos := max_pos;
    end;

    ret := 0;
    for j := 0 to pos do ret := max(ret, a[id][j]);
    writeln(ret);
  end;
  close(input); close(output);
end.

File: 2019/Thang03_2019/16_03_2019/dance.cpp
#include <iostream>
#include <stdio.h>

using namespace std;
//////

const
    int maxn=210, oo=2e9+100;
int T, n;
int l[maxn], p[maxn];
int f[2][maxn*maxn];
//////

void enter()
{
    scanf("%d\n", &n);
    for (int i=1; i<=n; i++) scanf("%d", &l[i]);
    scanf("\n");
    for (int i=1; i<=n; i++) scanf("%d", &p[i]);
    scanf("\n");
}
//////

void process()
{
    int min_pos, max_pos=0, pos, id=0;
    f[0][0]=0;
    for (int i=1; i<=n; i++)
    {
        id=1-id;
        pos=max_pos;
        if (l[i]>0) max_pos+=l[i];

        for (int j=0; j<=pos; j++) f[id][j]=f[1-id][j];
        for (int j=pos+1; j<=max_pos; j++) f[id][j]=-oo;

        min_pos=pos+l[i];
        for (int j=max(0, l[i]); j<=min_pos; j++)
            f[id][j]=max(f[id][j], f[1-id][j-l[i]]+p[i]);
    }

    int res=-oo;
    for (int j=0; j<=max_pos; j++)
        res=max(res, f[id][j]);
    cout << res << '\n';
}
//////

int main()
{
    freopen("dance.inp", "r", stdin);
    freopen("dance.out", "w", stdout);
    scanf("%d/n", &T);
    while (T--)
    {
        enter();
        process();
    }
}
